// babel-less/routes/batchAudit/auditTanbleComponent.jsx
import classNames from 'classnames';
import _ from 'lodash';
import Card from 'react-toolbox/lib/card/Card';
import React from 'react';
import { createSelector } from 'reselect';
import PropTypes from 'prop-types';
import { AgGridReact, reactCellRendererFactory } from 'ag-grid-react';
import {
  FieldGroup,
  Modal,
  Button,
  Checkbox,
  FormGroup,
  ControlLabel,
  FormControl,
} from 'react-bootstrap';
import formatDate from './../../shared/dateTime';
import { CAN_EDIT, FILTER } from '../../constants/stateConstants';
import AuditEntry from './../../model/auditEntry';

const defaultColDef = {
  headerClass: 'scenarios__header-cell',
  cellClass: 'scenarios__cell',
  suppressMenu: true,
  overlayLoadingTemplate:
    '<span class="ag-overlay-loading-center">Please wait...</span>',
};

const getColumnDefs = () => [
  {
    headerName: 'Batch Name',
    field: 'batch.batchName',
    cellClass: 'scenario-audit-dialog__name-cell',
    minWidth: 100,
    suppressSorting: true,
  },
  {
    headerName: 'Added',
    field: 'added',
    cellClass: 'scenario-audit-dialog__description-cell',
    minWidth: 200,
    suppressSorting: true,
  },
  {
    headerName: 'Removed',
    field: 'removed',
    cellClass: 'scenario-audit-dialog__updatedBy-cell',
    minWidth: 200,
    suppressSorting: true,
  },
  {
    headerName: 'Active Scenarios',
    field: 'scenarios',
    cellClass: 'scenario-audit-dialog__updated-cell',
    minWidth: 200,
  },
  {
    headerName: 'Reason for change',
    field: 'batch.deltaComment',
    cellClass: 'scenario-audit-dialog__updatedBy-cell',
    minWidth: 200,
    suppressSorting: true,
  },
  {
    headerName: 'Updated',
    field: 'timestamp',
    cellClass: 'scenario-audit-dialog__updated-cell',
    minWidth: 120,
    maxWidth: 150,
    cellRenderer: ({ value }) => formatDate(value),
  },
  {
    headerName: 'Updated By',
    field: 'updatedBy',
    cellClass: 'scenario-audit-dialog__updatedBy-cell',
    minWidth: 180,
  },
  {
    headerName: '',
    field: 'defaultCol',
    cellClass: 'scenarios__updatedBy-cell',
    minWidth: 10,
    maxWidth: 10,
  },
];

const sortModel = [
  {
    colId: 'timestamp',
    sort: 'desc',
  },
];

const adaptAuditEntry = audit => {
  let added = [];
  let removed = [];
  let scenarios = [];
  if (audit.batch != null) {
    added =
      audit.batch.added !== undefined
        ? audit.batch.added.map(a => `${a.scenarioName} (v.${a.fileVersion})`)
        : [];
    removed =
      audit.batch.removed !== undefined
        ? audit.batch.removed.map(a => `${a.scenarioName} (v.${a.fileVersion})`)
        : [];
    scenarios =
      audit.batch.scenarios !== undefined
        ? audit.batch.scenarios.map(
            a => `${a.scenarioName} (v.${a.fileVersion})`
          )
        : [];
  }
  return {
    updatedBy: audit.updatedBy,
    timestamp: audit.timestamp,
    batch: audit.batch,
    deltaComment: audit.deltaComment,
    added: added.toString(),
    removed: removed.toString(),
    scenarios: scenarios.toString(),
  };
};

class AuditTableComponent extends React.PureComponent {
  constructor() {
    super();

    this.initialState = {
      busy: true,
      isReady: false,
      selectedVersion: undefined,
      currentSelectedVersion: undefined,
      rowData: [],
    };
    this.state = this.initialState;

    this.onGridReady = this.onGridReady.bind(this);
    this.sizeColumnsToFit = _.debounce(this.sizeColumnsToFit.bind(this), 500);
    this.filterColumn = this.filterColumn.bind(this);
    this.exportToExcel = this.exportToExcel.bind(this);

    this.getColumnDefs = createSelector(
      props => props[CAN_EDIT],
      getColumnDefs
    );
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.batchList) {
      this.batchlist = nextProps.batchList;
      if (this.agGrid) {
        this.agGrid.api.showLoadingOverlay();
      }

      const batchfilter = nextProps.batchList.filter(
        b =>
          b.batchName ===
          (nextProps.batchName === '' ? b.batchName : nextProps.batchName)
      );

      const auditlist = batchfilter.map(y => y.auditList);

      let combinedLists = [];
      if (auditlist !== undefined && auditlist !== [] && auditlist.length > 0)
        combinedLists = auditlist.reduce((a, b) => [...a, ...b]);

      const auditEntryList = combinedLists
        .filter(f => f.batch !== undefined)
        .map(adaptAuditEntry);

      this.setState({
        isReady: true,
        busy: false,
        rowData: auditEntryList,
      });
    }

    if (nextProps[FILTER] || nextProps[FILTER] === '') {
      this.filterColumn(nextProps[FILTER]);
    }
  }

  onGridReady(agGrid) {
    this.agGrid = agGrid;

    this.state.rowData.forEach((row, index) => agGrid.api.sizeColumnsToFit());

    if (this.agGrid) agGrid.api.setSortModel(sortModel);

    this.sizeColumnsToFit(() => this.setState({ isReady: true, busy: false }));
  }

  sizeColumnsToFit(callback) {
    if (this.agGrid) {
      this.agGrid.api.sizeColumnsToFit();
    }

    if (_.isFunction(callback)) {
      callback();
    }
  }

  filterColumn(filterVal) {
    if (!this.agGrid) return;

    this.agGrid.api.setQuickFilter(filterVal);
  }

  exportToExcel() {
    if (!this.agGrid) return;

    const params = {
      skipHeader: false,
      columnGroups: true,
      skipFooters: true,
      skipGroups: true,
      skipPinnedTop: true,
      skipPinnedBottom: true,
      allColumns: true,
      onlySelected: true,
      suppressQuotes: true,
      fileName: `Batch_Audit_${Date.now()}.csv`,
      columnSeparator: ',',
    };

    this.agGrid.api.exportDataAsCsv(params);
  }

  render() {
    const { busy, isReady, rowData } = this.state;

    const columnDefs = this.getColumnDefs(this);

    return (
      <div
        className={classNames(
          'ag-grid ag-fresh sc-grid',
          this.props.busy ? 'busy' : '',
          this.props.className
        )}
        style={{ height: '90%' }}
      >
        <Card
          className={classNames(
            'ag-fresh ag-grid sc-grid',
            'scenarios__card',
            { busy },
            {
              'scenarios__card--hidden': !isReady,
            }
          )}
        >
          <div className="scenarios__grid-container" style={{ height: '100%' }}>
            <AgGridReact
              columnDefs={columnDefs}
              defaultColDef={defaultColDef}
              enableColResize
              editType="fullRow"
              enableFilter={false}
              enableSorting
              onGridReady={this.onGridReady}
              onAfterSortChanged={this.onAfterSortChanged}
              onRowClicked={this.onRowClicked}
              onSelectionChanged={this.onSelectionChanged}
              rowClass="scenario-audit-dialog__row"
              rowData={rowData}
              rowSelection="single"
              suppressContextMenu
              suppressMovableColumns
              suppressRowClickSelection
            />
          </div>
        </Card>
      </div>
    );
  }
}

AuditTableComponent.propTypes = {
  batchName: PropTypes.string,
  batchList: PropTypes.arrayOf(PropTypes.object),
};

AuditTableComponent.defaultProps = {
  batchName: '',
  batchList: [],
};

export default AuditTableComponent;
