// babel-less/routes/batchAudit/batchAuditContainer.jsx
import _ from 'lodash';
import React, { PropTypes } from 'react';
import {
  Button,
  Col,
  Grid,
  Row,
  FormGroup,
  FormControl,
  ControlLabel,
  Glyphicon,
  DropdownButton,
  MenuItem,
  InputGroup,
} from 'react-bootstrap';
import classNames from 'classnames';
import {
  BATCH_LIST,
  CAN_EDIT,
  SCENARIO_LIST,
  FILTER,
} from '../../constants/stateConstants';
import { REQUEST_BATCHES } from '../../constants/actionTypes';
import Batch from '../../model/batch';
import formatDate from '../../shared/dateTime';
import ListComponent, { VALUE } from '../../shared/components/listComponent';
import AuditTableComponent from './auditTableComponent';
import GlyphButton from '../../shared/components/glyphButton';

class BatchAuditContainer extends React.PureComponent {
  constructor(props) {
    super(props);

    this.selectBatch = this.selectBatch.bind(this);

    this.state = {
      batchName: '',
      busy: true,
    };
  }

  componentWillMount() {
    this.props[REQUEST_BATCHES]();
  }

  selectBatch(batch) {
    this.setState({
      batchName: batch,
    });
  }

  batchSelector() {
    return (
      <DropdownButton
        componentClass={InputGroup.Button}
        bsSize="small"
        id="input-dropdown-addon"
        title={this.state.batchName !== '' ? this.state.batchName : 'All'}
      >
        <MenuItem key="1" selected onSelect={() => this.selectBatch('')}>
          All
        </MenuItem>
        {this.props[BATCH_LIST].map((s, idx) => {
          /* eslint-disable react/no-array-index-key */
          if (this.state.batchName === s.batchName) {
            return (
              <MenuItem
                key={idx}
                onSelect={() => this.selectBatch(s.batchName)}
                active
              >
                {s.batchName}
              </MenuItem>
            );
          }
          return (
            <MenuItem key={idx} onSelect={() => this.selectBatch(s.batchName)}>
              {s.batchName}
            </MenuItem>
          );
        })}
      </DropdownButton>
    );
  }

  render() {
    const { busy } = this.props;

    return (
      <div
        className={classNames('scenarioContainer', 'batchAuditContainer', {
          busy,
        })}
      >
        <div>
          <h1>Batch Audit</h1>
        </div>
        <div style={{ display: 'flex' }}>
          <div id="container">
            <div className="box">
              <h3 key={1}>Active batch:</h3>
            </div>
            <div className="box">{this.batchSelector()}</div>
          </div>
          <div className="alignRight">
            <GlyphButton
              onCellClicked={() => {
                this.AuditTableComponent.exportToExcel();
              }}
              iconName="download-alt"
              className="scenarioAuditContainer__download-cell"
              tooltipText="Download as CSV"
            />
          </div>
        </div>
        <div style={{ height: '90%' }}>
          <AuditTableComponent
            ref={instance => {
              this.AuditTableComponent = instance;
            }}
            batchName={this.state.batchName}
            batchList={this.props[BATCH_LIST]}
            {...{ [FILTER]: this.props[FILTER] }}
          />
        </div>
      </div>
    );
  }
}

BatchAuditContainer.PropTypes = {
  [BATCH_LIST]: PropTypes.arrayOf(PropTypes.object).isRequired,
  busy: PropTypes.string,
  [FILTER]: PropTypes.string.isRequired,
};

export default BatchAuditContainer;
