// babel-less/routes/batch/deltaDialog.jsx
import classNames from 'classnames';
import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { Button, Modal } from 'react-bootstrap';
import DeltaListComponent from './deltaListComponent';

class DeltaDialog extends React.Component {
  constructor() {
    super();

    this.onClose = this.onClose.bind(this);
  }

  onClose() {
    this.props.onClose();
    this.setState({ isReady: false });
  }

  render() {
    const { active, delta } = this.props;

    return (
      <Modal
        className={classNames('delta-dialog')}
        show={active}
        onHide={this.onClose}
      >
        <Modal.Header closeButton>
          <Modal.Title>Delta</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <div className="delta-dialog__content">
            {delta.added.length > 0 || delta.removed.length > 0 ? (
              <DeltaListComponent
                dataList={delta.combined}
                id={'id'}
                header={''}
                value={'scenarioNameAndVersion'}
                added={delta.added}
                removed={delta.removed}
              />
            ) : null}
          </div>
        </Modal.Body>
        <Modal.Footer>
          {' '}
          <Button value="OK" bsStyle="default" onClick={this.onClose}>
            OK
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

DeltaDialog.propTypes = {
  active: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  delta: PropTypes.shape({
    added: PropTypes.arrayOf(PropTypes.object),
    removed: PropTypes.arrayOf(PropTypes.object),
    combined: PropTypes.arrayOf(PropTypes.object),
  }),
};

DeltaDialog.defaultProps = { delta: { added: [], removed: [], combined: [] } };

export default DeltaDialog;
