// babel-less/batchManagementDescriptor.js
import React from 'react';
import scenarioRoute from './routes/scenario/scenarioIndex';
import batchRoute from './routes/batch/batchIndex';
import batchAuditRoute from './routes/batchAudit/batchAuditIndex';
import scenarioAuditRoute from './routes/scenarioAudit/scenarioAuditIndex';
import BatchManagementSidebar from './components/batchManagementSidebar';
import ContentContainer from '../shell/components/contentContainer';
import getBatchesAndScenariosEpic from './epics/getBatchesAndScenariosEpic';
import saveBatchEpics from './epics/saveBatchEpic';
import batchReducer from './reducers/batchReducer';
import getScenariosEpic from './epics/getScenariosEpic';
import saveScenariosEpic from './epics/saveScenariosEpic';
import setScenarioVersionEpic from './epics/setScenarioVersionEpic';
import getScenarioVersionsEpic from './epics/getScenarioVersionsEpic';
import archiveScenariosEpic from './epics/archiveScenariosEpic';
import getScenariosAuditEpic from './epics/getScenariosAuditEpic';
import scenarioReducer from './reducers/scenarioReducer';
import batchManagementConfigurationReducer from './reducers/batchManagementConfigurationReducer';
import { BATCHES, CONFIGURATION, SCENARIOS } from './constants/reducerTypes';
import Entitlement from './../auth/model/entitlement';

const reducers = {
  [SCENARIOS]: scenarioReducer,
  [BATCHES]: batchReducer,
  [CONFIGURATION]: batchManagementConfigurationReducer,
};

const epics = [
  getBatchesAndScenariosEpic,
  saveBatchEpics,
  getScenariosEpic,
  saveScenariosEpic,
  archiveScenariosEpic,
  getScenarioVersionsEpic,
  setScenarioVersionEpic,
  getScenariosAuditEpic,
];

const route = [
  {
    path: 'batch-management',
    components: {
      content: ContentContainer,
      sidebar: BatchManagementSidebar,
    },
    indexRoute: {
      onEnter: ({ location: { pathname } }, replace) => {
        replace('batch-management/batch');
      },
    },
    childRoutes: [
      {
        path: 'scenario',
        component: scenarioRoute,
      },
      {
        path: 'batch',
        component: batchRoute,
      },
      {
        path: 'batch-audit',
        component: batchAuditRoute,
      },
      {
        path: 'scenario-audit',
        component: scenarioAuditRoute,
      },
    ],
  },
];

const bootstrapActions = [];

const metadata = {
  supportedEnvironments: ['all'],
  menuConfig: {
    title: 'Batch Management',
    icon: 'fa-clock-o',
    tooltip: 'Batch Management',
    pathname: '/batch-management',
  },
  authorisation: {
    entitlement: [
      Entitlement.from('ui.batch-management', 'read-only'),
      Entitlement.from('ui.batch-management', 'read-write'),
    ],
    role: [],
  },
};

export default { reducers, epics, route, bootstrapActions, metadata };
