// babel-less/shared/components/listComponent.js
import PropTypes from 'prop-types';
import React from 'react';
import { Tooltip, OverlayTrigger } from 'react-bootstrap';

import { CSSTransitionGroup } from 'react-transition-group';

export const VALUE = 'value';
export const COMPONENT = 'component';

const formatLongDisplay = value => {
  if (value === '' || value === undefined) return null;

  let result = <span>{value}</span>;
  const tooltip = <Tooltip id="tooltip">{value}</Tooltip>;

  if (value.length > 40) {
    result = (
      <OverlayTrigger placement="top" overlay={tooltip}>
        <span>{value.substring(0, 40)}...</span>
      </OverlayTrigger>
    );
  }
  return result;
};

const ListComponent = ({
  dataList,
  dataListType,
  header,
  subHeader,
  selectedItem,
  id,
  value,
  onClick,
  onDoubleClick,
}) => (
  <div>
    <span>{header && header.length ? <h3> {header} </h3> : null}</span>
    <ul className="list-group bm-list-group">
      <CSSTransitionGroup
        transitionName="bm-list-group"
        transitionEnterTimeout={500}
        transitionLeaveTimeout={300}
      >
        {dataList.map(
          (s, idx) => (
            /* eslint-disable react/no-array-index-key */
            <li
              key={idx}
              className={
                selectedItem && s[id] === selectedItem[id] ? (
                  'list-group-item active'
                ) : (
                  'list-group-item'
                )
              }
              onClick={onClick ? () => onClick(s) : null}
              onDoubleClick={onDoubleClick ? () => onDoubleClick(s) : null}
            >
              {dataListType === VALUE ? (
                formatLongDisplay(s[value])
              ) : (
                formatLongDisplay(s)
              )}
            </li>
          )
          /* eslint-enable react/no-array-index-key */
        )}
      </CSSTransitionGroup>
    </ul>
    {subHeader && subHeader.length ? <h5> {subHeader} </h5> : null}
  </div>
);

ListComponent.propTypes = {
  dataList: PropTypes.arrayOf(PropTypes.object).isRequired,
  dataListType: PropTypes.oneOf([VALUE, COMPONENT]),
  // eslint-disable-next-line react/forbid-prop-types
  selectedItem: PropTypes.object,
  header: PropTypes.string.isRequired,
  subHeader: PropTypes.string,
  id: PropTypes.string,
  value: PropTypes.string,

  onClick: PropTypes.func,
  onDoubleClick: PropTypes.func,
};

ListComponent.defaultProps = {
  dataListType: VALUE,
  header: undefined,
  selectedItem: undefined,
  subHeader: undefined,
  id: undefined,
  value: undefined,

  onClick: undefined,
  onDoubleClick: undefined,
};

export default ListComponent;
